File: Java/src/main/java/com/gurujadhav/cacheclient/TCPClient.java
package com.gurujadhav.cacheclient;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

/**
 * Simple TCP socket client wrapper supporting blocking send and frame-aware receive.
 */
public class TCPClient {
    private final String domain;
    private final int port;
    private Socket socket;
    private InputStream inputStream;
    private OutputStream outputStream;

    public TCPClient(String domain, int port) {
        this.domain = domain;
        this.port = port;
    }

    /**
     * Resolves target address and connects to CacheCore server.
     * @return boolean - true if connection was established
     */
    public boolean connect() {
        try {
            InetAddress[] addresses = InetAddress.getAllByName(domain);
            for (InetAddress addr : addresses) {
                try {
                    socket = new Socket();
                    // Set a connection timeout of 5 seconds
                    socket.connect(new InetSocketAddress(addr, port), 5000);
                    inputStream = socket.getInputStream();
                    outputStream = socket.getOutputStream();
                    return true;
                } catch (IOException e) {
                    close();
                }
            }
        } catch (IOException e) {
            throw new NetworkException("DNS resolution failed for: " + domain, e);
        }
        throw new NetworkException("failed to connect to: " + domain + ":" + port);
    }

    /**
     * Sends raw bytes over TCP and blocks until the full RESP frame response is received.
     * @return String - Raw RESP response string
     */
    public synchronized String send(String raw) {
        if (socket == null || !socket.isConnected() || socket.isClosed()) {
            throw new NetworkException("client is not connected to CacheCore");
        }
        try {
            byte[] bytes = raw.getBytes(StandardCharsets.UTF_8);
            outputStream.write(bytes);
            outputStream.flush();
            return recv();
        } catch (IOException e) {
            throw new NetworkException("network error while sending to CacheCore", e);
        }
    }

    private String recv() {
        byte[] buffer = new byte[4096];
        StringBuilder response = new StringBuilder();

        try {
            while (true) {
                int bytesRead = inputStream.read(buffer);
                if (bytesRead == -1) {
                    throw new NetworkException("server closed the connection");
                }

                response.append(new String(buffer, 0, bytesRead, StandardCharsets.UTF_8));

                // Verify completeness of the RESP response stream using RESPParser
                if (RESPParser.isComplete(response.toString())) {
                    return response.toString();
                }
            }
        } catch (IOException e) {
            throw new NetworkException("network error while receiving from CacheCore", e);
        }
    }

    /**
     * Closes socket and stream resources.
     */
    public void close() {
        try {
            if (inputStream != null) inputStream.close();
            if (outputStream != null) outputStream.close();
            if (socket != null) socket.close();
        } catch (IOException ignored) {}
    }
}

File: CPP/src/TCPClient.cpp
#include "../include/TCPClient.h"
#include "../include/Exceptions.h"
#include "../include/RESPParser.h"
#include <sys/socket.h>
#include <sys/types.h>
#include <netdb.h>
#include <unistd.h>

TCPClient::TCPClient(const std::string &_domain, const std::uint16_t _port)
    : domain(_domain), port(_port) {}

bool TCPClient::connect() {
    return tryConnect();
}

bool TCPClient::tryConnect() {
    struct addrinfo hints{};
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;

    struct addrinfo *res = nullptr;

    int status = getaddrinfo(domain.c_str(), std::to_string(port).c_str(), &hints, &res);
    if (status != 0) {
        throw NetworkException("DNS resolution failed for: " + domain);
    }

    for (struct addrinfo *p = res; p != nullptr; p = p->ai_next) {
        clientFd = socket(p->ai_family, p->ai_socktype, p->ai_protocol);
        if (clientFd < 0)
            continue;
        if (::connect(clientFd, p->ai_addr, p->ai_addrlen) == 0) {
            freeaddrinfo(res);
            return true;
        }
        ::close(clientFd);
        clientFd = -1;
    }

    freeaddrinfo(res);
    throw NetworkException("failed to connect to: " + domain + ":" + std::to_string(port));
}

std::string TCPClient::RECV() {
    char buffer[4096];
    std::string response;

    while (true) {
        int bytesRead = recv(clientFd, buffer, sizeof(buffer) - 1, 0);

        if (bytesRead == 0) {
            throw NetworkException("server closed the connection");
        }
        if (bytesRead == -1) {
            throw NetworkException("network error while receiving from CacheCore");
        }

        buffer[bytesRead] = '\0';
        response += std::string(buffer, bytesRead);

        if (RESPParser::isComplete(response)) {
            return response;
        }
    }
}

std::string TCPClient::SEND(const std::string &raw) {
    size_t totalSent = 0;
    while (totalSent < raw.size()) {
        ssize_t bytesSent = send(clientFd, raw.c_str() + totalSent, raw.size() - totalSent, 0);
        if (bytesSent == -1) {
            throw NetworkException("network error while sending to CacheCore");
        }
        totalSent += bytesSent;
    }
    return RECV();
}

TCPClient::~TCPClient() {
    close();
}

void TCPClient::close() {
    if (clientFd != -1) {
        ::close(clientFd);
        clientFd = -1;
    }
}
